pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -320,12 +320,20 @@ func (s *shardActiveSeriesMiddleware) mergeResponses(ctx context.Context, respon
 			}
 
 			for it.ReadArray() {
-				item := labelBuilderPool.Get().(*labels.Builder)
-				it.ReadMapCB(func(iterator *jsoniter.Iterator, s string) bool {
-					item.Set(s, iterator.ReadString())
-					return true
-				})
-				items <- item
+				select {
```

Comment:
i think you could save the `select` statement and simply check against `ctx.Err` at the beginning of each iteration. according to the documentation ([ref](https://github.com/golang/go/blob/master/src/context/context.go#L108)), a non-nil value is always expected when the `Done` channel is closed.

Reply:
I'm a bit worried about the performance implications of calling `ctx.Err()` on every iteration, what do you think about the periodic check I added in 2040eb8?

Reply:
Sorry for a post-merge comment, but wanted to note on this one:

> I'm a bit worried about the performance implications of calling `ctx.Err()` on every iteration

We actually have benchmarks for this code path, which we can execute with

```
% go test ./pkg/frontend/querymiddleware/ -run XX -bench 'BenchmarkActiveSeriesMiddlewareMergeResponses/encoding=none'
```

When one compares the overhead of an `if` inside the loop and allocating a `ticker` per response, the former adds a much lower overhead (+1% -vs- +7%) ✌️ 

```
% benchstat ~/tmp/{1,2,3}.txt                                                                                                                                                            ⋯
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/frontend/querymiddleware
                                                                     │ /Users/v/tmp/1.txt │         /Users/v/tmp/2.txt         │         /Users/v/tmp/3.txt         │
                                                                     │       sec/op       │   sec/op     vs base               │   sec/op     vs base               │
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-2-4          7.400µ ± 0%   7.911µ ± 1%  +6.91% (p=0.000 n=10)   7.482µ ± 0%  +1.11% (p=0.000 n=10)

                                                                     │ /Users/v/tmp/1.txt │          /Users/v/tmp/2.txt          │         /Users/v/tmp/3.txt          │
                                                                     │        B/op        │     B/op      vs base                │     B/op      vs base               │
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-2-4         1.539Ki ± 0%   1.977Ki ± 0%  +28.46% (p=0.000 n=10)   1.586Ki ± 0%  +3.05% (p=0.000 n=10)

                                                                     │ /Users/v/tmp/1.txt │         /Users/v/tmp/2.txt         │       /Users/v/tmp/3.txt       │
                                                                     │     allocs/op      │ allocs/op   vs base                │ allocs/op   vs base            │
ActiveSeriesMiddlewareMergeResponses/encoding=none/num-responses-2-4           34.00 ± 0%   40.00 ± 0%  +17.65% (p=0.000 n=10)   34.00 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```

Note, `1.txt` are the results from the original code; `2.txt` the merged version with ticker; `3.txt` a version with `if ctx.Err() != nil` inside the loop.

Reply:
Thanks, see follow up in https://github.com/grafana/mimir/pull/7396